File: Services/DashboardServer.cs
using System.Net;
using System.Security.Cryptography;
using System.Text;
using System.Threading.RateLimiting;
using EpikChatBot.Core;

namespace EpikChatBot.Services;

public sealed class DashboardServer(
    BotConfig config,
    Func<string?, Task<object>> getStatus,
    Func<string, string, string, Task<object>> setSetting,
    Func<string, string, Task<object>> runAction) : IAsyncDisposable
{
  private const long MaximumRequestBytes = 16 * 1024;
  private WebApplication? _app;

  public async Task StartAsync(CancellationToken cancellationToken)
  {
    if (!config.DashboardEnabled)
    {
      return;
    }

    if (Encoding.UTF8.GetByteCount(config.DashboardToken) < 32)
    {
      throw new InvalidOperationException("DASHBOARD_TOKEN must contain at least 32 UTF-8 bytes when the dashboard is enabled.");
    }

    var host = LoopbackHost(config.DashboardHost);
    if (!host.Equals(config.DashboardHost, StringComparison.OrdinalIgnoreCase))
    {
      Console.Error.WriteLine($"[dashboard] Refusing non-loopback host '{config.DashboardHost}'; binding to {host}. Use an HTTPS reverse proxy for remote access.");
    }

    var builder = WebApplication.CreateSlimBuilder();
    builder.WebHost.UseUrls($"http://{host}:{config.DashboardPort}");
    builder.Services.AddRateLimiter(options =>
    {
      options.RejectionStatusCode = StatusCodes.Status429TooManyRequests;
      options.GlobalLimiter = PartitionedRateLimiter.Create<HttpContext, string>(context =>
              RateLimitPartition.GetFixedWindowLimiter(
                  context.Connection.RemoteIpAddress?.ToString() ?? "local",
                  _ => new FixedWindowRateLimiterOptions
                  {
                    PermitLimit = 60,
                    QueueLimit = 0,
                    Window = TimeSpan.FromMinutes(1),
                  }));
    });

    _app = builder.Build();
    _app.Use(async (context, next) =>
    {
      context.Response.Headers.XContentTypeOptions = "nosniff";
      context.Response.Headers.XFrameOptions = "DENY";
      context.Response.Headers["Referrer-Policy"] = "no-referrer";
      context.Response.Headers.ContentSecurityPolicy = "default-src 'self'; script-src 'unsafe-inline'; style-src 'unsafe-inline'; base-uri 'none'; frame-ancestors 'none'";
      if (context.Request.ContentLength > MaximumRequestBytes)
      {
        context.Response.StatusCode = StatusCodes.Status413PayloadTooLarge;
        return;
      }

      await next();
    });
    _app.UseRateLimiter();

    _app.MapGet("/health", () => Results.Json(new { status = "ok" }));
    _app.MapGet("/", () => Results.Content(Html, "text/html", Encoding.UTF8));
    _app.MapGet("/api/status", async (HttpRequest request) =>
        Authorized(request) ? Results.Json(await getStatus(request.Query["room"])) : Results.Unauthorized());
    _app.MapPost("/api/settings", async (HttpRequest request) =>
    {
      if (!Authorized(request))
      {
        return Results.Unauthorized();
      }

      var body = await request.ReadFromJsonAsync<SettingRequest>(cancellationToken: cancellationToken);
      return body is null ? Results.BadRequest() : Results.Json(await setSetting(body.RoomId, body.Path, body.Value));
    });
    _app.MapPost("/api/action", async (HttpRequest request) =>
    {
      if (!Authorized(request))
      {
        return Results.Unauthorized();
      }

      var body = await request.ReadFromJsonAsync<ActionRequest>(cancellationToken: cancellationToken);
      return body is null ? Results.BadRequest() : Results.Json(await runAction(body.RoomId, body.Action));
    });

    await _app.StartAsync(cancellationToken);
    Console.WriteLine($"[dashboard] http://{host}:{config.DashboardPort}");
  }

  internal static string LoopbackHost(string host)
  {
    if (host.Equals("localhost", StringComparison.OrdinalIgnoreCase))
    {
      return "localhost";
    }

    return IPAddress.TryParse(host, out var address) && IPAddress.IsLoopback(address)
        ? host
        : "127.0.0.1";
  }

  private bool Authorized(HttpRequest request)
  {
    const string bearer = "Bearer ";
    var header = request.Headers.Authorization.ToString();
    if (!header.StartsWith(bearer, StringComparison.OrdinalIgnoreCase))
    {
      return false;
    }

    var supplied = Encoding.UTF8.GetBytes(header[bearer.Length..]);
    var expected = Encoding.UTF8.GetBytes(config.DashboardToken);
    return supplied.Length == expected.Length && CryptographicOperations.FixedTimeEquals(supplied, expected);
  }

  public async ValueTask DisposeAsync()
  {
    if (_app is not null)
    {
      await _app.StopAsync();
      await _app.DisposeAsync();
    }
  }

  private sealed record SettingRequest(string RoomId, string Path, string Value);

  private sealed record ActionRequest(string RoomId, string Action);

  private const string Html = """<!doctype html><html><head><meta charset="utf-8"><meta name="viewport" content="width=device-width"><title>EpikChat Bot</title><style>body{font:16px system-ui;max-width:900px;margin:2rem auto;padding:0 1rem;background:#10151d;color:#e8eef7}button,input,select{font:inherit;padding:.5rem;margin:.25rem}pre{background:#1b2430;padding:1rem;overflow:auto;border-radius:.5rem}</style></head><body><h1>EpikChat Bot</h1><input id="token" type="password" placeholder="Dashboard token"><input id="room" placeholder="Room ID"><button onclick="refresh()">Refresh</button><pre id="status">Enter token and refresh.</pre><script>async function refresh(){const r=await fetch('/api/status?room='+encodeURIComponent(room.value),{headers:{authorization:'Bearer '+token.value}});status.textContent=JSON.stringify(await r.json(),null,2)}</script></body></html>""";
}

File: legacy-node/lib/dashboard-server.js
const http = require("http");
const https = require("https");
const fs = require("fs");
const { URL } = require("url");

const DASHBOARD_HTML = `<!doctype html>
<html><head><meta charset="utf-8"><meta name="viewport" content="width=device-width"><title>EpikChat Bot</title>
<style>body{font:16px system-ui;max-width:900px;margin:2rem auto;padding:0 1rem;background:#10151d;color:#e8eef7}button,input,select{font:inherit;padding:.5rem;margin:.25rem}pre{background:#1b2430;padding:1rem;overflow:auto;border-radius:.5rem}.row{display:flex;gap:.5rem;flex-wrap:wrap}</style></head>
<body><h1>EpikChat Bot</h1><div class="row"><input id="username" placeholder="Username"><input id="password" type="password" placeholder="Password"><button onclick="login()">Sign in</button><input id="token" type="password" placeholder="Legacy token"><input id="room" placeholder="Room ID"><button onclick="refresh()">Refresh</button></div>
<pre id="status">Enter token and refresh.</pre><h2>Action</h2><div class="row"><select id="action"><option>trivia-start</option><option>trivia-stop</option><option>marbles-open</option><option>marbles-close</option></select><button onclick="act()">Run</button></div>
<h2>Setting</h2><div class="row"><input id="path" placeholder="ai.enabled"><input id="value" placeholder="true"><button onclick="setting()">Save</button></div>
<script>let csrf='';const headers=()=>({'authorization':token.value?'Bearer '+token.value:'','x-csrf-token':csrf,'content-type':'application/json'});async function login(){const r=await fetch('/api/login',{method:'POST',headers:{'content-type':'application/json'},body:JSON.stringify({username:username.value,password:password.value})});const d=await r.json();csrf=d.csrf||'';password.value='';status.textContent=JSON.stringify(d,null,2);if(r.ok)refresh()}async function refresh(){const r=await fetch('/api/status?room='+encodeURIComponent(room.value),{headers:headers()});status.textContent=JSON.stringify(await r.json(),null,2)}async function act(){await fetch('/api/action',{method:'POST',headers:headers(),body:JSON.stringify({roomId:room.value,action:action.value})});refresh()}async function setting(){await fetch('/api/settings',{method:'POST',headers:headers(),body:JSON.stringify({roomId:room.value,path:path.value,value:value.value})});refresh()}</script></body></html>`;

function readBody(request) {
  return new Promise((resolve, reject) => {
    let body = "";
    request.on("data", (chunk) => {
      body += chunk;
      if (body.length > 1_000_000) reject(new Error("Request body too large"));
    });
    request.on("end", () => {
      try { resolve(body ? JSON.parse(body) : {}); } catch (error) { reject(error); }
    });
    request.on("error", reject);
  });
}

class DashboardServer {
  constructor({ host = "127.0.0.1", port = 8787, token, auth = null, tlsKeyFile = null, tlsCertFile = null, getStatus, setSetting, runAction, logger = console }) {
    this.host = host;
    this.port = port;
    this.token = token;
    this.auth = auth;
    this.tlsKeyFile = tlsKeyFile; this.tlsCertFile = tlsCertFile;
    this.getStatus = getStatus;
    this.setSetting = setSetting;
    this.runAction = runAction;
    this.logger = logger;
    this.server = null;
  }

  sessionToken(request) { return String(request.headers.cookie || "").split(/;\s*/).find((item) => item.startsWith("epik_session="))?.slice(13) || ""; }

  authorization(request) {
    if (this.token && request.headers.authorization === `Bearer ${this.token}`) return { role: "owner", method: "token" };
    const token = this.sessionToken(request); const session = this.auth?.authenticate(token);
    return session ? { ...session, token, method: "session" } : null;
  }

  authorized(request, role = "viewer") {
    const auth = this.authorization(request); const levels = { viewer: 0, owner: 1 };
    return auth && levels[auth.role] >= levels[role] ? auth : null;
  }

  json(response, status, data) {
    response.writeHead(status, { "content-type": "application/json; charset=utf-8", "cache-control": "no-store", "x-content-type-options": "nosniff", "content-security-policy": "default-src 'none'; frame-ancestors 'none'" });
    response.end(JSON.stringify(data));
  }

  async handle(request, response) {
    const url = new URL(request.url, `http://${request.headers.host || "localhost"}`);
    if (url.pathname === "/health") {
      const status = await this.getStatus(null);
      return this.json(response, status.ok ? 200 : 503, status);
    }
    if (url.pathname === "/" && request.method === "GET") {
      response.writeHead(200, { "content-type": "text/html; charset=utf-8", "x-frame-options": "DENY", "x-content-type-options": "nosniff", "content-security-policy": "default-src 'self'; script-src 'unsafe-inline'; style-src 'unsafe-inline'; frame-ancestors 'none'" });
      response.end(DASHBOARD_HTML);
      return;
    }
    if (url.pathname === "/api/login" && request.method === "POST" && this.auth) {
      const body = await readBody(request); const session = this.auth.login(body.username, body.password, request.socket.remoteAddress);
      if (!session) return this.json(response, 401, { error: "invalid-credentials" });
      response.setHeader("set-cookie", `epik_session=${session.token}; HttpOnly; SameSite=Strict; Path=/; Max-Age=${Math.floor((session.expiresAt - Date.now()) / 1000)}${this.tlsKeyFile ? "; Secure" : ""}`);
      return this.json(response, 200, { ok: true, username: session.username, role: session.role, csrf: session.csrf });
    }
    const authorization = this.authorized(request, url.pathname === "/api/status" ? "viewer" : "owner");
    if (!authorization) return this.json(response, 401, { error: "unauthorized" });
    if (request.method !== "GET" && authorization.method === "session" && request.headers["x-csrf-token"] !== authorization.csrf) return this.json(response, 403, { error: "csrf" });
    if (url.pathname === "/api/accounts" && request.method === "GET") return this.json(response, 200, { accounts: this.auth?.listAccounts() || [] });
    if (url.pathname === "/api/accounts" && request.method === "POST" && this.auth) {
      const body = await readBody(request); return this.json(response, 201, this.auth.createAccount(body.username, body.password, body.role === "owner" ? "owner" : "viewer"));
    }
    if (url.pathname === "/api/logout" && request.method === "POST" && this.auth) { this.auth.logout(authorization.token); response.setHeader("set-cookie", "epik_session=; HttpOnly; SameSite=Strict; Path=/; Max-Age=0"); return this.json(response, 200, { ok: true }); }
    if (url.pathname === "/api/status" && request.method === "GET") {
      return this.json(response, 200, await this.getStatus(url.searchParams.get("room")));
    }
    if (url.pathname === "/api/settings" && request.method === "POST") {
      const body = await readBody(request);
      return this.json(response, 200, await this.setSetting(body));
    }
    if (url.pathname === "/api/action" && request.method === "POST") {
      const body = await readBody(request);
      return this.json(response, 200, await this.runAction(body));
    }
    return this.json(response, 404, { error: "not-found" });
  }

  start() {
    if (this.server) return;
    const listener = (request, response) => {
      this.handle(request, response).catch((error) => {
        this.logger.error("[dashboard]", error);
        if (!response.headersSent) this.json(response, 500, { error: "internal-error" });
        else response.end();
      });
    };
    this.server = this.tlsKeyFile && this.tlsCertFile
      ? https.createServer({ key: fs.readFileSync(this.tlsKeyFile), cert: fs.readFileSync(this.tlsCertFile) }, listener)
      : http.createServer(listener);
    const protocol = this.tlsKeyFile && this.tlsCertFile ? "https" : "http";
    this.server.listen(this.port, this.host, () => this.logger.log(`[dashboard] ${protocol}://${this.host}:${this.port}`));
  }

  stop() {
    return new Promise((resolve) => {
      if (!this.server) return resolve();
      this.server.close(() => resolve());
      this.server = null;
    });
  }
}

module.exports = { DashboardServer };
